File: electron/sqlite/queries/invoices.ts
/**
 * 发票 & 收款记录 SQLite 查询模块
 *
 * 实现 invoices、payment_records 两张表的 CRUD 操作。
 * 包含发票状态同步逻辑。
 */

import log from 'electron-log'
import { tryGetSqlite, rowToCamel, toSqliteValue } from './helpers'

// ═══════════════════════════════════════════════════════════════════════════════
// Invoices — 列映射
// ═══════════════════════════════════════════════════════════════════════════════

const INV_COLUMNS: Record<string, string> = {
  id: 'id',
  type: 'type',
  status: 'status',
  invoiceKind: 'invoice_kind',
  invoiceNo: 'invoice_no',
  invoiceCode: 'invoice_code',
  name: 'name',
  amount: 'amount',
  taxAmount: 'tax_amount',
  priceAmount: 'price_amount',
  taxRate: 'tax_rate',
  issueDate: 'issue_date',
  sellerId: 'seller_id',
  buyerId: 'buyer_id',
  settlementId: 'settlement_id',
  projectId: 'project_id',
  contractId: 'contract_id',
  receivedAmount: 'received_amount',
  fileUrl: 'file_url',
  fileType: 'file_type',
  remarks: 'remarks',
  createdAt: 'created_at',
  updatedAt: 'updated_at',
}

const INV_INSERT_COLS = Object.values(INV_COLUMNS).filter(c => c !== 'id')
const INV_INSERT_SQL = `INSERT INTO invoices (${INV_INSERT_COLS.map(c => `"${c}"`).join(', ')}) VALUES (${INV_INSERT_COLS.map(() => '?').join(', ')})`

// ═══════════════════════════════════════════════════════════════════════════════
// Payment Records — 列映射
// ═══════════════════════════════════════════════════════════════════════════════

const PAY_COLUMNS: Record<string, string> = {
  id: 'id',
  type: 'type',
  amount: 'amount',
  recordDate: 'record_date',
  projectId: 'project_id',
  partnerId: 'partner_id',
  contractId: 'contract_id',
  invoiceDetails: 'invoice_details',
  remarks: 'remarks',
  createdAt: 'created_at',
  fileUrl: 'file_url',
  fileType: 'file_type',
}

const PAY_INSERT_COLS = Object.values(PAY_COLUMNS).filter(c => c !== 'id')
const PAY_INSERT_SQL = `INSERT INTO payment_records (${PAY_INSERT_COLS.map(c => `"${c}"`).join(', ')}) VALUES (${PAY_INSERT_COLS.map(() => '?').join(', ')})`

// ═══════════════════════════════════════════════════════════════════════════════
// 辅助
// ═══════════════════════════════════════════════════════════════════════════════

function toInsertParams(columns: Record<string, string>, insertCols: string[], obj: Record<string, any>): any[] {
  return insertCols.map(col => {
    const jsonKey = Object.entries(columns).find(([, c]) => c === col)?.[0]
    if (!jsonKey) return null
    return toSqliteValue(obj[jsonKey])
  })
}

function toUpdateSet(columns: Record<string, string>, changes: Record<string, any>, excludeKeys: string[] = []): { sql: string; params: any[] } {
  const setClauses: string[] = []
  const params: any[] = []
  for (const [jsonKey, value] of Object.entries(changes)) {
    if (excludeKeys.includes(jsonKey)) continue
    const col = columns[jsonKey]
    if (!col) continue
    setClauses.push(`"${col}" = ?`)
    params.push(toSqliteValue(value))
  }
  return { sql: setClauses.join(', '), params }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Invoices — 操作
// ═══════════════════════════════════════════════════════════════════════════════

/** 列出发票（含 sellerName、buyerName、projectName、contractName、computed receivedAmount/status） */
export function listInvoices(type?: string): any[] | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    let sql = `
      SELECT i.*,
        s.name as seller_name,
        b.name as buyer_name,
        p.name as project_name,
        COALESCE(ic.name, ec.name, ac.name, '') as contract_name
      FROM invoices i
      LEFT JOIN partners s ON i.seller_id = s.id
      LEFT JOIN partners b ON i.buyer_id = b.id
      LEFT JOIN projects p ON i.project_id = p.id
      LEFT JOIN income_contracts ic ON i.contract_id = ic.id
      LEFT JOIN expense_contracts ec ON i.contract_id = ec.id
      LEFT JOIN agreement_contracts ac ON i.contract_id = ac.id
    `
    const params: any[] = []
    if (type) {
      sql += ' WHERE i.type = ?'
      params.push(type)
    }
    sql += ' ORDER BY i.created_at DESC'

    const rows = sqlite.prepare(sql).all(...params) as Record<string, any>[]
    return rows.map(row => {
      const camel = rowToCamel(row)
      camel.sellerName = (row as any).seller_name || ''
      camel.buyerName = (row as any).buyer_name || ''
      camel.projectName = (row as any).project_name || ''
      camel.contractName = (row as any).contract_name || ''
      delete (camel as any).seller_name
      delete (camel as any).buyer_name
      delete (camel as any).project_name
      delete (camel as any).contract_name
      return camel
    })
  } catch (err) {
    log.error('[SQLite] invoices.list error:', err)
    return null
  }
}

/** 创建发票 */
export function createInvoice(invoice: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const params = toInsertParams(INV_COLUMNS, INV_INSERT_COLS, invoice)
    sqlite.prepare(INV_INSERT_SQL).run(...params)
    return true
  } catch (err) {
    log.error('[SQLite] invoices.create error:', err)
    return false
  }
}

/** 更新发票 */
export function updateInvoice(invoice: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const { sql: setSql, params: setParams } = toUpdateSet(INV_COLUMNS, invoice, ['id'])
    if (!setSql) return true
    setParams.push(new Date().toISOString())
    setParams.push(invoice.id)
    const result = sqlite.prepare(`UPDATE invoices SET ${setSql}, "updated_at" = ? WHERE id = ?`).run(...setParams)
    return result.changes > 0
  } catch (err) {
    log.error('[SQLite] invoices.update error:', err)
    return false
  }
}

/** 更新发票状态 */
export function updateInvoiceStatus(id: number, status: string): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('UPDATE invoices SET status = ?, updated_at = ? WHERE id = ?')
      .run(status, new Date().toISOString(), id)
    return true
  } catch (err) {
    log.error('[SQLite] invoices.updateStatus error:', err)
    return false
  }
}

/** 更新发票的 receivedAmount 和 status */
export function updateInvoiceReceived(id: number, receivedAmount: number, status: string): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('UPDATE invoices SET received_amount = ?, status = ?, updated_at = ? WHERE id = ?')
      .run(receivedAmount, status, new Date().toISOString(), id)
    return true
  } catch (err) {
    log.error('[SQLite] invoices.updateReceived error:', err)
    return false
  }
}

/** 删除发票 */
export function deleteInvoice(id: number): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('DELETE FROM invoices WHERE id = ?').run(id)
    return true
  } catch (err) {
    log.error('[SQLite] invoices.delete error:', err)
    return false
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Payment Records — 操作
// ═══════════════════════════════════════════════════════════════════════════════

/** 列出收款记录（含 projectName、partnerName、contractName、invoiceInfos） */
export function listPaymentRecords(type?: string): any[] | null {
  const sqlite = tryGetSqlite()
  if (!sqlite) return null
  try {
    let sql = `
      SELECT pr.*,
        p.name as project_name,
        pt.name as partner_name,
        COALESCE(ic.name, ec.name, ac.name, '') as contract_name
      FROM payment_records pr
      LEFT JOIN projects p ON pr.project_id = p.id
      LEFT JOIN partners pt ON pr.partner_id = pt.id
      LEFT JOIN income_contracts ic ON pr.contract_id = ic.id
      LEFT JOIN expense_contracts ec ON pr.contract_id = ec.id
      LEFT JOIN agreement_contracts ac ON pr.contract_id = ac.id
    `
    const params: any[] = []
    if (type) {
      sql += ' WHERE pr.type = ?'
      params.push(type)
    }
    sql += ' ORDER BY pr.record_date DESC, pr.created_at DESC'

    const rows = sqlite.prepare(sql).all(...params) as Record<string, any>[]
    return rows.map(row => {
      const camel = rowToCamel(row)
      camel.projectName = (row as any).project_name || ''
      camel.partnerName = (row as any).partner_name || ''
      camel.contractName = (row as any).contract_name || ''
      delete (camel as any).project_name
      delete (camel as any).partner_name
      delete (camel as any).contract_name

      // 解析 invoiceDetails 并补充发票信息
      const invoiceDetails = camel.invoiceDetails || []
      const invoiceInfos = invoiceDetails.map((d: any) => {
        const invRow = sqlite.prepare('SELECT invoice_no, name, amount FROM invoices WHERE id = ?').get(d.invoiceId) as any
        return {
          invoiceId: d.invoiceId,
          invoiceNo: invRow?.invoice_no || '',
          invoiceName: invRow?.name || '',
          invoiceAmount: invRow?.amount || 0,
          paymentAmount: d.paymentAmount,
        }
      })
      camel.invoiceInfos = invoiceInfos
      return camel
    })
  } catch (err) {
    log.error('[SQLite] paymentRecords.list error:', err)
    return null
  }
}

/** 创建收款记录 */
export function createPaymentRecord(record: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const params = toInsertParams(PAY_COLUMNS, PAY_INSERT_COLS, record)
    sqlite.prepare(PAY_INSERT_SQL).run(...params)
    return true
  } catch (err) {
    log.error('[SQLite] paymentRecords.create error:', err)
    return false
  }
}

/** 更新收款记录 */
export function updatePaymentRecord(record: any): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const { sql: setSql, params: setParams } = toUpdateSet(PAY_COLUMNS, record, ['id'])
    if (!setSql) return true
    setParams.push(record.id)
    const result = sqlite.prepare(`UPDATE payment_records SET ${setSql} WHERE id = ?`).run(...setParams)
    return result.changes > 0
  } catch (err) {
    log.error('[SQLite] paymentRecords.update error:', err)
    return false
  }
}

/** 删除收款记录 */
export function deletePaymentRecord(id: number): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    sqlite.prepare('DELETE FROM payment_records WHERE id = ?').run(id)
    return true
  } catch (err) {
    log.error('[SQLite] paymentRecords.delete error:', err)
    return false
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// 发票状态重算（SQLite 版，供 paymentRecords 的 update/delete 后调用）
// ═══════════════════════════════════════════════════════════════════════════════

/** 重新计算所有发票的 receivedAmount 和 status */
export function recalculateInvoiceStatusSqlite(): boolean {
  const sqlite = tryGetSqlite()
  if (!sqlite) return false
  try {
    const invoices = sqlite.prepare('SELECT id, amount, received_amount FROM invoices').all() as { id: number; amount: number; received_amount: number }[]
    for (const inv of invoices) {
      // 从 payment_records 的 invoice_details JSON 中计算每张发票的累计收款
      const paymentRows = sqlite.prepare('SELECT invoice_details FROM payment_records').all() as { invoice_details: string }[]
      let totalReceived = 0
      for (const pr of paymentRows) {
        try {
          const details = JSON.parse(pr.invoice_details || '[]')
          for (const d of details) {
            if (d.invoiceId === inv.id) {
              totalReceived += (d.paymentAmount || 0)
            }
          }
        } catch { /* skip malformed JSON */ }
      }
      let status = 'issued'
      if (totalReceived >= (inv.amount || 0)) {
        status = 'received'
      } else if (totalReceived > 0) {
        status = 'partially_paid'
      }
      sqlite.prepare('UPDATE invoices SET received_amount = ?, status = ?, updated_at = ? WHERE id = ?')
        .run(totalReceived, status, new Date().toISOString(), inv.id)
    }
    return true
  } catch (err) {
    log.error('[SQLite] recalculateInvoiceStatus error:', err)
    return false
  }
}

File: src-tauri/src/commands/invoices.rs
/**
 * 发票管理命令
 *
 * 对应 Electron 版本的 invoices.ts
 */

use crate::db::AppState;
use crate::error::{AppError, AppResult};
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{command, State};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Invoice {
    pub id: i64,
    #[serde(rename = "type")]
    pub invoice_type: String, // invoice_in | invoice_out
    pub status: Option<String>,
    pub invoice_kind: Option<String>,
    pub invoice_no: Option<String>,
    pub invoice_code: Option<String>,
    pub name: Option<String>,
    pub amount: f64,
    pub tax_amount: f64,
    pub price_amount: f64,
    pub tax_rate: f64,
    pub issue_date: Option<String>,
    pub seller_id: Option<i64>,
    pub buyer_id: Option<i64>,
    pub settlement_id: Option<i64>,
    pub project_id: Option<i64>,
    pub contract_id: Option<i64>,
    pub received_amount: f64,
    pub file_url: Option<String>,
    pub file_type: Option<String>,
    pub remarks: Option<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub seller_name: Option<String>,
    pub buyer_name: Option<String>,
    pub project_name: Option<String>,
}

/// 获取发票列表
#[command]
pub fn get_invoices(
    state: State<'_, AppState>,
    project_id: Option<i64>,
    invoice_type: Option<String>,
) -> AppResult<Vec<Invoice>> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let mut sql = String::from(
        "SELECT i.id, i.type, i.status, i.invoice_kind, i.invoice_no, i.invoice_code,
                i.name, i.amount, i.tax_amount, i.price_amount, i.tax_rate, i.issue_date,
                i.seller_id, i.buyer_id, i.settlement_id, i.project_id, i.contract_id,
                i.received_amount, i.file_url, i.file_type, i.remarks, i.created_at, i.updated_at,
                s.name as seller_name,
                b.name as buyer_name,
                p.name as project_name
         FROM invoices i
         LEFT JOIN partners s ON i.seller_id = s.id
         LEFT JOIN partners b ON i.buyer_id = b.id
         LEFT JOIN projects p ON i.project_id = p.id
         WHERE 1=1"
    );

    let mut param_values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();
    let mut param_idx = 1;

    if let Some(pid) = project_id {
        sql.push_str(&format!(" AND i.project_id = ?{}", param_idx));
        param_values.push(Box::new(pid));
        param_idx += 1;
    }

    if let Some(ref itype) = invoice_type {
        sql.push_str(&format!(" AND i.type = ?{}", param_idx));
        param_values.push(Box::new(itype.clone()));
    }

    sql.push_str(" ORDER BY i.created_at DESC");

    let mut stmt = db.prepare(&sql).map_err(|e| AppError::Database(e.to_string()))?;

    let param_refs: Vec<&dyn rusqlite::types::ToSql> = param_values.iter().map(|p| p.as_ref()).collect();

    let invoices = stmt
        .query_map(param_refs.as_slice(), |row| {
            Ok(Invoice {
                id: row.get(0)?,
                invoice_type: row.get(1)?,
                status: row.get(2)?,
                invoice_kind: row.get(3)?,
                invoice_no: row.get(4)?,
                invoice_code: row.get(5)?,
                name: row.get(6)?,
                amount: row.get(7)?,
                tax_amount: row.get(8)?,
                price_amount: row.get(9)?,
                tax_rate: row.get(10)?,
                issue_date: row.get(11)?,
                seller_id: row.get(12)?,
                buyer_id: row.get(13)?,
                settlement_id: row.get(14)?,
                project_id: row.get(15)?,
                contract_id: row.get(16)?,
                received_amount: row.get(17)?,
                file_url: row.get(18)?,
                file_type: row.get(19)?,
                remarks: row.get(20)?,
                created_at: row.get(21)?,
                updated_at: row.get(22)?,
                seller_name: row.get(23)?,
                buyer_name: row.get(24)?,
                project_name: row.get(25)?,
            })
        })
        .map_err(|e| AppError::Database(e.to_string()))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::Database(e.to_string()))?;

    Ok(invoices)
}

/// 创建发票
#[command]
pub fn create_invoice(state: State<'_, AppState>, invoice: Invoice) -> AppResult<i64> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    db.execute(
        "INSERT INTO invoices (type, status, invoice_kind, invoice_no, invoice_code, name,
         amount, tax_amount, price_amount, tax_rate, issue_date, seller_id, buyer_id,
         settlement_id, project_id, contract_id, remarks)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17)",
        params![
            invoice.invoice_type,
            invoice.status,
            invoice.invoice_kind,
            invoice.invoice_no,
            invoice.invoice_code,
            invoice.name,
            invoice.amount,
            invoice.tax_amount,
            invoice.price_amount,
            invoice.tax_rate,
            invoice.issue_date,
            invoice.seller_id,
            invoice.buyer_id,
            invoice.settlement_id,
            invoice.project_id,
            invoice.contract_id,
            invoice.remarks,
        ],
    )
    .map_err(|e| AppError::Database(e.to_string()))?;

    let id = db.last_insert_rowid();
    Ok(id)
}

/// 更新发票
#[command]
pub fn update_invoice(state: State<'_, AppState>, invoice: Invoice) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE invoices SET type = ?1, status = ?2, invoice_kind = ?3, invoice_no = ?4,
             invoice_code = ?5, name = ?6, amount = ?7, tax_amount = ?8, price_amount = ?9,
             tax_rate = ?10, issue_date = ?11, seller_id = ?12, buyer_id = ?13,
             settlement_id = ?14, project_id = ?15, contract_id = ?16, remarks = ?17,
             updated_at = datetime('now')
             WHERE id = ?18",
            params![
                invoice.invoice_type,
                invoice.status,
                invoice.invoice_kind,
                invoice.invoice_no,
                invoice.invoice_code,
                invoice.name,
                invoice.amount,
                invoice.tax_amount,
                invoice.price_amount,
                invoice.tax_rate,
                invoice.issue_date,
                invoice.seller_id,
                invoice.buyer_id,
                invoice.settlement_id,
                invoice.project_id,
                invoice.contract_id,
                invoice.remarks,
                invoice.id,
            ],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("发票 {} 不存在", invoice.id)));
    }

    Ok(())
}

/// 删除发票
#[command]
pub fn delete_invoice(state: State<'_, AppState>, id: i64) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute("DELETE FROM invoices WHERE id = ?1", params![id])
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("发票 {} 不存在", id)));
    }

    Ok(())
}

/// 更新发票状态
#[command]
pub fn update_invoice_status(
    state: State<'_, AppState>,
    id: i64,
    status: String,
) -> AppResult<()> {
    let db = state.db.lock().map_err(|e| AppError::Database(e.to_string()))?;

    let affected = db
        .execute(
            "UPDATE invoices SET status = ?1, updated_at = datetime('now') WHERE id = ?2",
            params![status, id],
        )
        .map_err(|e| AppError::Database(e.to_string()))?;

    if affected == 0 {
        return Err(AppError::NotFound(format!("发票 {} 不存在", id)));
    }

    Ok(())
}

// ============ 收付款记录 ============

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PaymentRecord {
    pub id: i64,
    #[serde(rename = "type")]
    pub payment_type: String,
    pub amount: f64,
    pub record_date: Option<String>,
    pub project_id: Option<i64>,
    pub partner_id: Option<i64>,
    pub contract_id: Option<i64>,
    pub invoice_details: Option<String>, // JSON string
    pub remarks: Option<String>,
    pub file_url: Option<String>,
    pub file_type: Option<String>,
    pub created_at: Option<String>,
    pub project_name: Option<String>,
    pub partner_name: Option<String>,
}
